### ISSUE TITLE
Leaking open portals crash transactions on CockroachDB 

### ISSUE THREAD

**Issue by lbialy** [slick/slick#2144]
**DESCRIPTION:**

Hello, I have been working recently on using CockroachDB as the database for Akka-Persistence & Akka-Projection with both of them using Slick as the dba layer. I was able to verify that most of the stuff works out of the box due to the postgres wire protocol compatibility of the Cockroach, but I found that one of the features is broken due to the difference in resource management in Cockroach in comparison to PostgreSQL. Specifically - Postgres allows for portals (open datasets / cursors basically) to a) be interleaved on the same connection (you can open more then one dataset concurrently on the same connection), b) be left open without explicit closing and then COMMIT or ROLLBACK messages close them implicitly. Both of these assumptions are not true with CockroachDB - it allows for a single portal to be opened at the time and it has to be explicitly closed for Cockroach to allow opening another. Also, if you don't close the portal and issue a COMMIT or ROLLBACK message it will crash with an error. Slick implementation appears to use the assumption that portals can be multiplexed and also will be closed implicitly.

Further reading: 

https://github.com/cockroachdb/cockroach/issues/4035 < comments from jordanlewis and knz about Cockroach-specific handling of the psql wire protocol

https://github.com/cockroachdb/cockroach/issues/42912 < ticket regarding the need for explicit closing of portals

**STEPS TO REPRODUCE:**

I have built a small exploratory project that uses dockerized cockroach and compares the behaviour of raw jdbc connection, jdbc connection managed by hikari and Slick. The only dependencies are sbt and docker: https://github.com/VirtusLab/slick-cockroachdb-issue

You can clone it and run `sbt test`. You can also clone the PR branch (https://github.com/VirtusLab/slick/tree/CockroachDB-support-fix) and use sbt shell to publish it locally:
```
+2.13.4
publishLocal
```
and then modify the build.sbt file of slick-cockroachdb-issue to see that it resolves the cockroach issue.

The issue itself happens in scenarios like this:
```
    for {
      maybeExisting <- findById(id).map(_.forUpdate).result.headOption
      row = maybeExisting.map(cs => cs.copy(currentSum = cs.currentSum + number)).getOrElse(CurrentSum(id, number))
      _ <- currentSums.insertOrUpdate(row)
    } yield ()
```

I have found this due to akka-projection interface for handlers that uses Slick. It demands that a DBIO is returned and then executes several of them in batch transactionally. This triggers the failing scenario. From JDBC perspective the problem happens only when all of these happen together:
a) transaction is started (`conn.setAutoCommit(false)` is called as in RawJdbcIntegrationTest.scala:15)
b) the query limit is set to non-zero value (`stmt.setMaxRows(1)` is called as in RawJdbcTestOps.scala:12) 

`setMaxRows` is only used in Slick (as far as I can see) with `headOption` operator.

The problem can be resolved (and is resolved by PR attached to this issue) by adding explicit portal/resultset handling which forces JDBC postgres driver to explicitly close the portal on protocol level. This requires that fetchSize is set (for headOption it can be set to 1 as we never expect more than 1 row in the resultset as it sets maxRows to 1) and that ResultSet is explicitly closed *before* the Statement is closed. This is missing from Slick and should be handled in slick.jdbc.StatementInvoker on line 53 (in close method of anonymous PositionedResult instance as `rs.close()` before `st.close()`). 

The fetchSize can be provided manually by appending `.withStatementParameters(fetchSize = 1)` to the problematic DBIO, but given the fact that the only place in slick that provides a value different to zero for maxRows in a call to slick.jdbc.StatementInvoker.results method is the Invoker.foreach used in Invoker.firstOption:
```
  final def firstOption(implicit session: JdbcBackend#Session): Option[R] = {
    var res: Option[R] = None
    foreach({ x => res = Some(x) }, 1)
    res
  }

  final def foreach(f: R => Unit, maxRows: Int = 0)(implicit session: JdbcBackend#Session): Unit = {
    val it = iteratorTo(maxRows)
    try { it.foreach(f) } finally { it.close() }
  }
```
I have taken the liberty of adding a specific handling for a case of maxRows=1 (it basically sets fetchSize to 1 too as it seems like a very small change that possibly won't backfire).

**EXPECTED RESULT:**

I'd expect that transaction succeeds.

**ACTUAL RESULT:**

```
org.postgresql.util.PSQLException: ERROR: unimplemented: multiple active portals not supported
  Hint: You have attempted to use a feature that is not yet implemented.
See: https://github.com/cockroachdb/cockroach/issues/40195
  Location: File: distsql_running.go, Routine: init, Line: 775
  Server SQLState: 0A000
	at org.postgresql.core.v3.QueryExecutorImpl.receiveErrorResponse(QueryExecutorImpl.java:2553)
	at org.postgresql.core.v3.QueryExecutorImpl.processResults(QueryExecutorImpl.java:2285)
	at org.postgresql.core.v3.QueryExecutorImpl.execute(QueryExecutorImpl.java:323)
	at org.postgresql.jdbc.PgStatement.executeInternal(PgStatement.java:473)
	at org.postgresql.jdbc.PgStatement.execute(PgStatement.java:393)
	at org.postgresql.jdbc.PgPreparedStatement.executeWithFlags(PgPreparedStatement.java:164)
	at org.postgresql.jdbc.PgPreparedStatement.executeUpdate(PgPreparedStatement.java:130)
	at com.zaxxer.hikari.pool.ProxyPreparedStatement.executeUpdate(ProxyPreparedStatement.java:61)
	at com.zaxxer.hikari.pool.HikariProxyPreparedStatement.executeUpdate(HikariProxyPreparedStatement.java)
	at slick.jdbc.JdbcActionComponent$InsertActionComposerImpl$InsertOrUpdateAction.$anonfun$nativeUpsert$1(JdbcActionComponent.scala:584)
	at slick.jdbc.JdbcBackend$SessionDef.withPreparedStatement(JdbcBackend.scala:425)
	at slick.jdbc.JdbcBackend$SessionDef.withPreparedStatement$(JdbcBackend.scala:420)
	at slick.jdbc.JdbcBackend$BaseSession.withPreparedStatement(JdbcBackend.scala:489)
	at slick.jdbc.JdbcActionComponent$InsertActionComposerImpl.preparedInsert(JdbcActionComponent.scala:511)
	at slick.jdbc.JdbcActionComponent$InsertActionComposerImpl$InsertOrUpdateAction.nativeUpsert(JdbcActionComponent.scala:581)
	at slick.jdbc.JdbcActionComponent$InsertActionComposerImpl$InsertOrUpdateAction.f$1(JdbcActionComponent.scala:564)
	at slick.jdbc.JdbcActionComponent$InsertActionComposerImpl$InsertOrUpdateAction.run(JdbcActionComponent.scala:569)
	at slick.jdbc.JdbcActionComponent$SimpleJdbcProfileAction.run(JdbcActionComponent.scala:28)
	at slick.jdbc.JdbcActionComponent$SimpleJdbcProfileAction.run(JdbcActionComponent.scala:25)
	at slick.basic.BasicBackend$DatabaseDef$$anon$3.liftedTree1$1(BasicBackend.scala:276)
	at slick.basic.BasicBackend$DatabaseDef$$anon$3.run(BasicBackend.scala:276)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1130)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:630)
	at java.base/java.lang.Thread.run(Thread.java:832)
```

**SETUP:**

https://github.com/VirtusLab/slick-cockroachdb-issue

**Comment by lbialy** [comment]
FYI: The `slick-cockroachdb-issue` repo does log things that happen on postgresql driver level. You can see the problem yourself on the protocol level by running all the tests and then looking into slick's log file in repo root and searching for `PortalSuspended` message. More about the pgsql wire protocol can be found here: https://www.postgresql.org/docs/12/protocol-flow.html - look for `PortalSuspended` there too.

**Comment by nafg** [comment]
Closed by #2145
